### DISCUSSION TITLE
How to change Material property of .GlTF files on Android

### DISCUSSION THREAD

**Discussion by murielK** [google/filament#4109] [Q&A] [answered]
Hello, 

I'm trying to change the following material property of  a .glTF; "metalic" and "roughness".

Here is my attempt code:

```
private fun populateScene(asset: FilamentAsset) {
        val renderableManager = engine.renderableManager
        var count = 0
        val popRenderables = { count = asset.popRenderables(readyRenderables); count != 0 }

        while (popRenderables()) {
            for (i in 0 until count) {

                val entity = readyRenderables[i]

                val renderableInstance = renderableManager.getInstance(entity)
                renderableManager.setScreenSpaceContactShadows(renderableInstance, true)
                renderableManager.setCastShadows(renderableInstance, true)

                //create Material
                val material = Material.Builder().build(engine)

                material.setDefaultParameter("matelic", 0f)
                material.setDefaultParameter("roughness", 0f)

                val materialInstance = material.createInstance()

                renderableManager.setMaterialInstanceAt(renderableInstance,0,materialInstance)
            }
            scene.addEntities(readyRenderables.take(count).toIntArray())
        }

        scene.addEntities(asset.lightEntities)
    }
```

Executing that method above will crash my application

Please, what I'm doing wrong here? Because the logs it generate are very difficult for me to decrypt  and debug it since I have no real experience with 3D nor filament itself.

`    public void setMaterialInstanceAt(@EntityInstance int i, @IntRange(from = 0) int primitiveIndex,
            @NonNull MaterialInstance materialInstance)`

Above is the method i should be using on renderableManager instance for applying a new material at run time, however what this "primitiveIndex" parameter means? or where should i get that?

is there any other better ways to create Materials instances? or to apply materials at run time on .gltf ?

Also sometime this line can also cause a crash `val material = Material.Builder().build(engine)`

Any code would be really helpful please.

**Comment by romainguy** [comment, marked as answer]
It crashes because you are creating an empty material. A `Material` needs to be built by loading a file compiled with `matc` (you can see examples of materials in our sample apps).

An entity with a renderable component (from `RenderManager`) is made of several "primitives" (think submeshes with different materials, for instance a sword would have a primitive for the blade made of metal, and another primitive for the hilt made of leather). The primitive index is simply which primitive you want to modify.

In the particular case of glTF instead of creating new materials you could simply *get* the existing materials and modify their properties.

**Reply by murielK** [reply]
Thank you for the great explanation. 

How do I get that primitiveIndex value? is there an API for that? or should I just go through the .gltf file and read it from there?

also `        while (popRenderables()) {
            for (i in 0 until count) {

                val entity = readyRenderables[i]

                val entityName = asset.getName(entity)
                Logger.DEFAULT.d("Entity: $entityName")
                val renderableInstance = renderableManager.getInstance(entity)
                renderableManager.setScreenSpaceContactShadows(renderableInstance, true)
                renderableManager.setCastShadows(renderableInstance, true)

                //get materials from Asset
                asset.materialInstances.forEach { materialInstance ->
                    Logger.DEFAULT.d("MaterialInstance: ${materialInstance.name}")

                    materialInstance.setParameter("matelic", 0f)
                    materialInstance.setParameter("roughness", 0f)
                    
                }

            }
            scene.addEntities(readyRenderables.take(count).toIntArray())
        }`

I got the existing materials and modified them but still no luck :/

**Reply by romainguy** [reply]
You are setting parameters that don't exist. You can see the names used by the glTF library here: https://github.com/google/filament/blob/main/libs/gltfio/materials/ubershader.mat.in

They follow the names of the official spec.

**Reply by murielK** [reply]
It worked! Thanks a lot @romainguy 👍 🥇
